### PR TITLE
Return correct HTTP header Forwarded IP address with multiple proxies.

### PR DESCRIPTION
**Pull request by mtrippled**

### CODE REVIEW

File: src/ripple/rpc/impl/Role.cpp [hunk 1]
```diff
@@ -108,14 +109,7 @@ requestInboundEndpoint (Resource::Manager& manager,
...
 {
-    auto it = request.find("X-Forwarded-For");
-    if (it != request.end())
-    {
-        return boost::beast::http::ext_list{
-            it->value()}.begin()->first;
-    }
-
-    it = request.find("Forwarded");
+    auto it = request.find(boost::beast::http::field::forwarded);
```
Comment (mellery451) [inline src/ripple/rpc/impl/Role.cpp]: you might want to `#include` the header that defines the fields, just for completeness.

File: src/ripple/rpc/impl/Role.cpp [hunk 2]
```diff
@@ -137,10 +131,20 @@ forwardedFor(http_request_type const& request)
             return {};
 
         found += forStr.size();
-        auto pos{it->value().find(';', forStr.size())};
+        auto pos = std::string(found).find(';');
```
Comment (mellery451) [inline src/ripple/rpc/impl/Role.cpp]: can you do this without the temporary string ?(e.g. using the former string view methods, but perhaps correcting the start offset). I don't this is safe and in-fact the asan build is flagging it.

File: src/ripple/rpc/impl/Role.cpp (hunk 1 above)
Comment (mtrippled) [reply src/ripple/rpc/impl/Role.cpp]: fixed

File: src/ripple/rpc/impl/Role.cpp (hunk 2 above)
Comment (mtrippled) [reply src/ripple/rpc/impl/Role.cpp]: fixed

File: src/ripple/rpc/impl/Role.cpp [hunk 3]
```diff
@@ -137,10 +131,20 @@ forwardedFor(http_request_type const& request)
...
         if (pos != boost::string_view::npos)
-            return {found, pos + 1};
-        return {found, it->value().size() - forStr.size()};
+        {
+            return *boost::beast::http::token_list(boost::string_view(
+                found, pos)).begin();
+        }
+        return *boost::beast::http::token_list(boost::string_view(
+            found, it->value().size() - forStr.size())).begin();
+    }
```
Comment (miguelportilla) [inline src/ripple/rpc/impl/Role.cpp]: An alternate to consider:

```
        auto const pos ([&]() {
            auto const pos {boost::string_view(
                found, it->value().end() - found).find(';')};
            if (pos == boost::string_view::npos)
                return it->value().size() - forStr.size();
            return pos;
        }());

        return *boost::beast::http::token_list(
            boost::string_view(found, pos)).begin();```

Comment (mtrippled) [reply src/ripple/rpc/impl/Role.cpp]: fixed

File: src/ripple/rpc/impl/Role.cpp [hunk 4]
```diff
@@ -137,10 +133,23 @@ forwardedFor(http_request_type const& request)
...
+        std::size_t const pos ([&]()
+        {
+            std::size_t const pos{boost::string_view(
+                found, it->value().end() - found).find(';')};
+            if (pos == boost::string_view::npos)
+                return it->value().size() - forStr.size();
+            return pos;
+        }());
+
+        return *boost::beast::http::token_list(
```
Comment (jwbusch) [inline src/ripple/rpc/impl/Role.cpp]: nit: does this container have a `front()` method? slightly clearer vs. dereferencing the begin() iterator

File: src/ripple/rpc/impl/Role.cpp [hunk 5]
```diff
@@ -137,10 +133,23 @@ forwardedFor(http_request_type const& request)
...
+            boost::string_view(found, pos)).begin();
+    }
+
+    it = request.find("X-Forwarded-For");
+    if (it != request.end())
+    {
+        return *boost::beast::http::token_list(it->value()).begin();
```
Comment (jwbusch) [inline src/ripple/rpc/impl/Role.cpp]: This used to be `beast::http::ext_list` and now it's `beast::http::token_list` ... was that part of the bug fix?

Comment (mtrippled) [reply src/ripple/rpc/impl/Role.cpp]: @jwbusch Yes, "ext_list" returns extensions which are basically key=value pairs, each separated by comma. "token_list" returns single items (tokens), also separated by comma.

The X-Forwarded-For HTTP header contains IP addresses separated by columns. I don't know why I used ext_list originally, but it was a mistake.

**Pull request closed without merging**